README.md
<p align="center">
<img src="./examples/favicon.ico" alt="mark text" width="100" height="100">
</p>

<h3 align="center">v-charts-multiple-y</h3>

<p align="center">
  <a href="https://npmjs.org/package/v-charts-multiple-y">
    <img src="http://img.shields.io/npm/dm/v-charts-multiple-y.svg" alt="NPM downloads">
  </a>
  <a href="https://www.npmjs.org/package/v-charts-multiple-y">
    <img src="https://img.shields.io/npm/v/v-charts-multiple-y.svg" alt="Npm package">
  </a>
  <a>
    <img src="https://img.shields.io/badge/language-javascript-yellow.svg" alt="Language">
  </a>
  <a>
    <img src="https://img.shields.io/badge/license-MIT-000000.svg" alt="License">
  </a>
</p>

<p align="center">
  <a href="https://v-charts.js.org/#/en/">
    Document
  </a>
  <span> | </span>
  <a href="https://codesandbox.io/s/z69myovqzx">
    Sample Project
  </a>
  <span> | </span>
  <a>
    English
  </a>
  <span> | </span>
  <a href="https://github.com/djkloop/v-charts/blob/master/README_CN.md">
    中文
  </a>
</p>

> Chart components based on Vue2.x and Echarts

# This magic change library copyright belongs to the original author
- Thanks to the author for providing such a good library to use echarts in vue
- This library only adds a multi-axis function for the line chart (if you do not need to use the source library)
- Copyright still belongs to the original author


## Features
- **Uniform data format:** Use an uniform data format that both convient for frontend and backend, and also easy to create and edit.
- **Simplified configuration:** With simplified configuration items, complex requirements can be easily implemented.
- **Simple customization:** Provide a variety of custom Echarts way, you can easily set the chart options.

## Support

Modern browsers and Internet Explorer 10+, include pc and mobile browser.

## Install

```
npm i v-charts-multiple-y echarts -S
```

## Start

```html
<template>
  <div>
    <ve-line :data="chartData"></ve-line>
  </div>
</template>

<script>
import VeLine from 'v-charts/lib/line.common'
export default {
  components: { VeLine },
  data () {
    return {
      chartData: {
        columns: ['date', 'age', 'balance', 'people'],
        rows: [
          { 'date': '01-01', 'age': 1231, 'balance': '999', 'people': '578' },
          { 'date': '01-02', 'age': 1223, 'balance': '781', 'people': '1213' },
          { 'date': '01-03', 'age': 2123, 'balance': '321', 'people': '213' },
          { 'date': '01-04', 'age': 4123, 'balance': '654', 'people': '2312' },
          { 'date': '01-05', 'age': 3123, 'balance': '972', 'people': '1233' },
          { 'date': '01-06', 'age': 7123, 'balance': '124', 'people': '2313' }
        ]
      }，
      /*
        yAxisSite
          - site[Array<string>]: Corresponds to columns
            Note: The first one of the array is the left number axis / the second one is the right first one, and the others are sorted one by one.
          - offset[Array<number>]: Corresponds to columns
            Note: The first one of the array is the left number axis / the second one is the right first one, and the others are sorted one by one.
        yAxisName Keep consistent with the official website
      */
      settings: {
        yAxisSite: {
          site: ['age', 'balance', 'people'],
          offset: [0, 0, 80]
        },
        yAxisName: ['age', 'balance', 'people']
      }
    }
  }
}
</script>
```

## Changelog

Detailed changes for each release are documented in the [release notes](https://github.com/ElemeFE/v-charts/releases) or [ChangeLog](./CHANGELOG.md).

## Contribution

Please make sure to read the [Contributing Guide](./CONTRIBUTING.md) before making a pull request.

## License

[MIT](http://opensource.org/licenses/MIT)


CHANGELOG.md
<a name="1.0.0"></a>
# [1.0.0](https://github.com/djkloop/v-charts/compare/v1.19.0...v1.0.0) (2019-07-22)


### Features

* change css cdn protocol ([f3c31e0](https://github.com/djkloop/v-charts/commit/f3c31e0))
* change docs cdn protocol ([92e7d5b](https://github.com/djkloop/v-charts/commit/92e7d5b))



<a name="1.19.0"></a>
# [1.19.0](https://github.com/ElemeFE/v-charts/compare/v1.18.0...v1.19.0) (2018-11-02)


### Bug Fixes

* **pie:** support multi pie item ([0a99bd2](https://github.com/ElemeFE/v-charts/commit/0a99bd2))



<a name="1.18.0"></a>
# [1.18.0](https://github.com/ElemeFE/v-charts/compare/v1.17.10...v1.18.0) (2018-10-15)


### Features

* **core:** add log prop to comp ([7abca27](https://github.com/ElemeFE/v-charts/commit/7abca27))
* Improve the configuration of the seriesMap and wave in the configuration settings of the water polo map ([3e82c6f](https://github.com/ElemeFE/v-charts/commit/3e82c6f))
* add wordcloud chart ([30ddf30](https://github.com/ElemeFE/v-charts/commit/30ddf30))
* add liquidfill chart ([391d62d](https://github.com/ElemeFE/v-charts/commit/391d62d))



<a name="1.17.10"></a>
## [1.17.10](https://github.com/ElemeFE/v-charts/compare/v1.17.9...v1.17.10) (2018-08-10)


### Features

* **map:** hide map of no value part ([b2b0f39](https://github.com/ElemeFE/v-charts/commit/b2b0f39))



<a name="1.17.9"></a>
## [1.17.9](https://github.com/ElemeFE/v-charts/compare/v1.17.8...v1.17.9) (2018-07-30)


### Bug Fixes

* **core:** load error while change data ([3517aa5](https://github.com/ElemeFE/v-charts/commit/3517aa5))



<a name="1.17.8"></a>
## [1.17.8](https://github.com/ElemeFE/v-charts/compare/v1.17.7...v1.17.8) (2018-07-11)


### Features

* **core:** add hooks params of options and Echarts ([e4ef7bd](https://github.com/ElemeFE/v-charts/commit/e4ef7bd))



<a name="1.17.7"></a>
## [1.17.7](https://github.com/ElemeFE/v-charts/compare/v1.17.6...v1.17.7) (2018-07-10)


### Bug Fixes

* **heatmap:** display error while data us empty ([9e6b16a](https://github.com/ElemeFE/v-charts/commit/9e6b16a))



<a name="1.17.6"></a>
## [1.17.6](https://github.com/ElemeFE/v-charts/compare/v1.17.5...v1.17.6) (2018-06-26)


### Features

* **core:** add notSetUnchange prop ([48d4346](https://github.com/ElemeFE/v-charts/commit/48d4346))



<a name="1.17.5"></a>
## [1.17.5](https://github.com/ElemeFE/v-charts/compare/v1.17.4...v1.17.5) (2018-06-25)


### Bug Fixes

* **line:** line display error while use dataZoom ([6cdf155](https://github.com/ElemeFE/v-charts/commit/6cdf155)), closes [#394](https://github.com/ElemeFE/v-charts/issues/394) [#387](https://github.com/ElemeFE/v-charts/issues/387)



<a name="1.17.4"></a>
## [1.17.4](https://github.com/ElemeFE/v-charts/compare/v1.17.3...v1.17.4) (2018-06-22)


### Features

* **core:** add resize checker ([5f8196d](https://github.com/ElemeFE/v-charts/commit/5f8196d))
* **style:** change mask status opacity and add blur ([7063c4a](https://github.com/ElemeFE/v-charts/commit/7063c4a))



<a name="1.17.3"></a>
## [1.17.3](https://github.com/ElemeFE/v-charts/compare/v1.17.2...v1.17.3) (2018-06-19)


### Features

* **funnel:** add `filterZero` and `useDefaultOrder` attribute in settings ([#373](https://github.com/ElemeFE/v-charts/issues/373)) ([687491d](https://github.com/ElemeFE/v-charts/commit/687491d))


test/index.js
/* eslint-disable no-undef */
import Vue from 'vue'
import chartData from '../examples/data/index.js'
import 'echarts/lib/component/tooltip'
import 'echarts/lib/component/legend'
import {
  VeLine,
  VeBar,
  VeHistogram,
  VePie,
  VeRing,
  VeFunnel,
  VeRadar,
  VeWaterfall,
  VeChart,
  VeMap,
  VeSankey,
  VeHeatmap,
  VeScatter,
  VeCandle,
  VeGauge,
  VeLiquidfill,
  VeWordcloud
} from '../lib/index.esm'

window.Promise = require('es6-promise').Promise

const comps = {
  line: VeLine,
  bar: VeBar,
  histogram: VeHistogram,
  pie: VePie,
  ring: VeRing,
  funnel: VeFunnel,
  radar: VeRadar,
  waterfall: VeWaterfall,
  chart: VeChart,
  map: VeMap,
  sankey: VeSankey,
  heatmap: VeHeatmap,
  scatter: VeScatter,
  candle: VeCandle,
  gauge: VeGauge,
  liquidfill: VeLiquidfill,
  wordcloud: VeWordcloud
}
let box
let vm = {}
createBox()

afterEach(() => {
  if (vm.$el) document.body.removeChild(vm.$el)
  createBox()
})

Object.keys(comps).forEach(type => {
  chartData[type].data.forEach(item => {
    describe(type + ': ', () => {
      testMount(type, comps[type], item)
    })
  })
})

function testMount (type, comp, item) {
  it(item.name, () => {
    const Ctor = Vue.extend(comp)
    const vm = new Ctor({
      propsData: { data: item.data, settings: item.settings, grid: item.grid }
    }).$mount(box)
    expect(vm.$el.classList.contains('ve-' + type)).toEqual(true)
  })
}

function createBox () {
  box = document.createElement('div')
  box.id = 'app'
  document.body.appendChild(box)
}


examples/data/line.js
export default {
  name: '折线图',
  type: 'line',
  data: [
    {
      name: '简单折线图',
      data: {
        columns: ['日期', '年龄', '余额', '人数'],
        rows: [
          { '日期': '1-1', '余额': '1213', '年龄': 3, '人数': 238 },
          { '日期': '1-2', '余额': 1223, '年龄': 6, '人数': 267 },
          { '日期': '1-3', '余额': 2123, '年龄': 9, '人数': 350 },
          { '日期': '1-4', '余额': 4123, '年龄': 12, '人数': 129 },
          { '日期': '1-5', '余额': 3123, '年龄': 15, '人数': 269 },
          { '日期': '1-6', '余额': 7123, '年龄': 20, '人数': 300 }
        ]
      }
    },
    {
      name: '多轴折线图',
      data: {
        columns: ['日期', '年龄', '余额', '人数'],
        rows: [
          { '日期': '1-1', '余额': '1213', '年龄': 3, '人数': 238 },
          { '日期': '1-2', '余额': 1223, '年龄': 6, '人数': 267 },
          { '日期': '1-3', '余额': 2123, '年龄': 9, '人数': 350 },
          { '日期': '1-4', '余额': 4123, '年龄': 12, '人数': 129 },
          { '日期': '1-5', '余额': 3123, '年龄': 15, '人数': 269 },
          { '日期': '1-6', '余额': 7123, '年龄': 20, '人数': 300 }
        ]
      },
      grid: {
        right: '25%'
      },
      settings: {
        yAxisSite: {
          site: ['年龄', '余额', '人数']
        },
        yAxisName: ['年龄', '余额', '人数']
      }
    },
    {
      name: '时间轴折线图',
      data: {
        columns: ['日期', '余额', '年龄'],
        rows: [
          { '日期': new Date(1512097200000), '余额': 123, '年龄': 3 },
          { '日期': new Date(1512097220000), '余额': 1223, '年龄': 6 },
          { '日期': new Date(1512097230000), '余额': 2123, '年龄': 9 },
          { '日期': new Date(1512097240000), '余额': 4123, '年龄': 12 },
          { '日期': new Date(1512097260000), '余额': 3123, '年龄': 15 },
          { '日期': new Date(1512097280000), '余额': 7123, '年龄': 20 }
        ]
      },
      settings: {
        xAxisType: 'time'
      }
    },
    {
      name: 'label 属性配置',
      data: {
        columns: ['日期', '余额', '年龄'],
        rows: [
          { '日期': '1-1', '余额': 123, '年龄': 3 },
          { '日期': '1-2', '余额': 1223, '年龄': 6 },
          { '日期': '1-3', '余额': 2123, '年龄': 9 },
          { '日期': '1-4', '余额': 4123, '年龄': 12 },
          { '日期': '1-5', '余额': 3123, '年龄': 15 },
          { '日期': '1-6', '余额': 7123, '年龄': 20 }
        ]
      },
      settings: {
        label: {
          normal: {
            show: true
          }
        }
      }
    },
    {
      name: '设置指标名称',
      data: {
        columns: ['date', 'balance', 'age'],
        rows: [
          { 'date': '1-1', 'balance': 123, 'age': 3 },
          { 'date': '1-2', 'balance': 1223, 'age': 6 },
          { 'date': '1-3', 'balance': 2123, 'age': 9 },
          { 'date': '1-4', 'balance': 4123, 'age': 12 },
          { 'date': '1-5', 'balance': 3123, 'age': 15 },
          { 'date': '1-6', 'balance': 7123, 'age': 20 }
        ]
      },
      settings: {
        labelMap: {
          date: '日期',
          balance: '余额'
        }
      }
    },
    {
      name: '设置legend名称',
      data: {
        columns: ['日期', '余额', '年龄'],
        rows: [
          { '日期': '1-1', '余额': 123, '年龄': 3 },
          { '日期': '1-2', '余额': 1223, '年龄': 6 },
          { '日期': '1-3', '余额': 2123, '年龄': 9 },
          { '日期': '1-4', '余额': 4123, '年龄': 12 },
          { '日期': '1-5', '余额': 3123, '年龄': 15 },
          { '日期': '1-6', '余额': 7123, '年龄': 20 }
        ]
      },
      settings: {
        legendName: {
          '余额': 'remain'
        }
      }
    },
    {
      name: '带有较小百分比数值',
      data: {
        columns: ['日期', '比率'],
        rows: [
          { '日期': '1-1', '余额': 123, '比率': 0.00001 },
          { '日期': '1-2', '余额': 1223, '比率': 0.00002 },
          { '日期': '1-3', '余额': 2123, '比率': 0.00003 },
          { '日期': '1-4', '余额': 4123, '比率': 0.00007 },
          { '日期': '1-5', '余额': 3123, '比率': 0.00001 },
          { '日期': '1-6', '余额': 7123, '比率': 0.00003 }
        ]
      },
      settings: {
        yAxisType: ['percent'],
        digit: 4
      }
    },
    {
      name: '坐标轴配置',
      data: {
        columns: ['日期', '余额', '比率'],
        rows: [
          { '日期': '1-1', '余额': 123, '比率': 0.3 },
          { '日期': '1-2', '余额': 1223, '比率': 0.6 },
          { '日期': '1-3', '余额': 2123, '比率': 0.9 },
          { '日期': '1-4', '余额': 4123, '比率': 0.12 },
          { '日期': '1-5', '余额': 3123, '比率': 0.15 },
          { '日期': '1-6', '余额': 7123, '比率': 0.20 }
        ]
      },
      settings: {
        axisSite: {
          right: ['比率']
        },
        yAxisType: ['KMB', 'percent'],
        yAxisName: ['数值', '比率']
      }
    },
    {
      name: '指标维度配置',
      data: {
        columns: ['日期', '余额', '比率'],
        rows: [
          { '日期': '1-1', '余额': 123, '比率': 0.1 },
          { '日期': '1-2', '余额': 1223, '比率': 0.2 },
          { '日期': '1-3', '余额': 2123, '比率': 0.3 },
          { '日期': '1-4', '余额': 4123, '比率': 0.4 },
          { '日期': '1-5', '余额': 3123, '比率': 0.5 },
          { '日期': '1-6', '余额': 7123, '比率': 0.6 }
        ]
      },
      settings: {
        dimension: ['比率'],
        metrics: ['余额']
      }
    },
    {
      name: '坐标轴缩放配置',
      data: {
        columns: ['日期', '余额', '比率'],
        rows: [
          { '日期': '1-1', '余额': 1232, '比率': 0.1 },
          { '日期': '1-2', '余额': 1223, '比率': 0.2 },
          { '日期': '1-3', '余额': 2123, '比率': 0.3 },
          { '日期': '1-4', '余额': 4123, '比率': 0.4 },
          { '日期': '1-5', '余额': 3123, '比率': 0.5 },
          { '日期': '1-6', '余额': 7123, '比率': 0.6 }
        ]
      },
      settings: {
        dimension: ['比率'],
        metrics: ['余额'],
        scale: [true, true]
      }
    },
    {
      name: '坐标轴值域配置',
      data: {
        columns: ['日期', '余额', '比率'],
        rows: [
          { '日期': '1-1', '余额': 1232, '比率': 0.1 },
          { '日期': '1-2', '余额': 1223, '比率': 0.2 },
          { '日期': '1-3', '余额': 2123, '比率': 0.3 },
          { '日期': '1-4', '余额': 4123, '比率': 0.4 },
          { '日期': '1-5', '余额': 3123, '比率': 0.5 },
          { '日期': '1-6', '余额': 7123, '比率': 0.6 }
        ]
      },
      settings: {
        dimension: ['比率'],
        metrics: ['余额'],
        min: [1000],
        max: [5000]
      }
    },
    {
      name: '面积图',
      data: {
        columns: ['日期', '余额', '比率'],
        rows: [
          { '日期': '1-1', '余额': 123, '比率': 0.1 },
          { '日期': '1-2', '余额': 1223, '比率': 0.2 },
          { '日期': '1-3', '余额': 2123, '比率': 0.3 },
          { '日期': '1-4', '余额': 4123, '比率': 0.4 },
          { '日期': '1-5', '余额': 3123, '比率': 0.5 },
          { '日期': '1-6', '余额': 7123, '比率': 0.6 }
        ]
      },
      settings: {
        area: true
      }
    },
    {
      name: '堆叠折线图',
      data: {
        columns: ['日期', '2015', '2016'],
        rows: [
          { '日期': '1-1', '2015': 100, '2016': 40 },
          { '日期': '1-2', '2015': 110, '2016': 60 },
          { '日期': '1-3', '2015': 140, '2016': 10 },
          { '日期': '1-4', '2015': 260, '2016': 30 },
          { '日期': '1-5', '2015': 300, '2016': 20 },
          { '日期': '1-6', '2015': 380, '2016': 70 }
        ]
      },
      settings: {
        stack: {
          '年份': ['2015', '2016']
        }
      }
    }
  ]
}
